src/components/FeatureSetsPanel/FeatureSetsPanelDataSource/FeatureSetsPanelDataSource.js

```diff
@@ -18,11 +40,203 @@ const FeatureSetsPanelDataSource = ({
 }) => {
   const [data, setData] = useState({
     kind: 'csv',
-    url: '',
+    url: {
+      pathType: '',
+      projectItemType: '',
+      project: '',
+      artifact: '',
+      placeholder: '',
+      path: '',
+      artifactReference: ''
+    },
     attributes: [],
     schedule: ''
   })
   const [showSchedule, setShowSchedule] = useState(false)
+  const [comboboxMatches, setComboboxMatches] = useState([])
+  const [projects, setProjects] = useState([])
+  const [artifacts, setArtifacts] = useState([])
+  const [artifactsReferences, setArtifactsReferences] = useState([])
+  const [urlProjectItemTypeEntered, setUrlProjectItemTypeEntered] = useState(
+    false
+  )
+  const [urlProjectPathEntered, setUrlProjectPathEntered] = useState(false)
+  const [urlArtifactPathEntered, setUrlArtifactPathEntered] = useState(false)
+  const [
+    urlArtifactReferencePathEntered,
+    setUrlArtifactReferencePathEntered
+  ] = useState(false)
+
+  useEffect(() => {
+    if (
+      data.url.pathType === MLRUN_STORAGE_INPUT_PATH_SCHEME &&
+      urlProjectItemTypeEntered &&
+      projects.length === 0
+    ) {
+      let projectsList = []
+
+      if (projectStore.projects.length === 0) {
+        fetchProjects().then(projects => {
+          projectsList = projects
+        })
+      } else {
+        projectsList = projectStore.projects
+      }
+
+      setProjects(generateProjectsList(projectsList, project))
```

Comment:
`fetchProjects` is an asynchronous function.
The control flow will execute it (initiate the API request) and _immediately_ continue to the next statement:
```js
setProjects(generateProjectsList(projectsList, project))
```
when `projectsList` equals `[]`.
Is this what you meant here?
When the response returns and `projectsList = projects` is executed, the `setProjects` will not be called again.
Did you mean to `await` the `fetchProjects` call, so the control flow _waits_ until the response is received?